File: python/src/apx_agent/_hot_swap.py
"""hot_swap_model — change a deployed agent's LLM without re-logging.

The runtime side (``_chat_agent.py::_resolve_model``) reads the
``APX_AGENT_MODEL_OVERRIDE`` env var and uses it instead of the
compile-time model. This module is the control side: writes the env
var onto a deployed serving endpoint so the next replica picks up the
new model.

Use cases:
  * Swap from ``claude-sonnet`` to ``claude-opus`` to test a more
    capable model without re-deploying the agent artifact.
  * Roll back a problematic model change in seconds (set the override
    to the previous model; agent picks it up on next call).
  * Run experiments where you'd otherwise re-log dozens of artifact
    versions just to tweak the LLM.

For traffic-split A/B testing across two complete artifacts (not just
the LLM endpoint), see ``apx_agent._canary`` instead.

Mechanism: ``ws.serving_endpoints.update_config_and_wait`` rewrites
each served entity's ``environment_vars`` to include the override.
Other env vars are preserved. The update is durable until the next
``hot_swap_model`` call or until the endpoint is redeployed.

Caveats:
  * The new env var takes effect on the next replica start. With
    scale-to-zero, the next request triggers a cold start that picks
    it up. Without scale-to-zero, existing replicas may serve the
    old model until they cycle.
  * Only affects agents that were compiled via apx-agent's
    ``compile_to_chat_agent`` / ``log_agent`` path. Hand-rolled agents
    that don't go through ``_chat_agent.py`` won't honor the override.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from databricks.sdk import WorkspaceClient

# The env var name the framework runtime reads in ``_chat_agent.py``.
# Exposed as a module constant so callers can match it in their own tooling.
APX_MODEL_OVERRIDE_ENV = "APX_AGENT_MODEL_OVERRIDE"


@dataclass(frozen=True)
class HotSwapResult:
    """Outcome of a hot-swap operation."""

    endpoint_name: str
    new_model: str
    previous_model: str | None  # the value the override held before; None if first set
    served_entities_updated: int


def hot_swap_model(
    endpoint_name: str,
    model: str,
    *,
    ws: "WorkspaceClient | None" = None,
    env_var: str = APX_MODEL_OVERRIDE_ENV,
    wait: bool = True,
) -> HotSwapResult:
    """Hot-swap a deployed agent's LLM endpoint by updating env vars.

    Args:
        endpoint_name: The serving endpoint hosting the agent (the name
            ``databricks.agents.deploy`` produced).
        model: The new model serving endpoint name (e.g.
            ``databricks-claude-opus-4-7``).
        ws: WorkspaceClient. Defaults to the framework's standard
            resolution (picks up DATABRICKS_HOST / OAuth M2M env).
        env_var: Override the env var name if you've customized the
            runtime variable. Defaults to ``APX_AGENT_MODEL_OVERRIDE``.
        wait: Block until the config update completes. Default True.

    Returns:
        ``HotSwapResult`` with the previous override value (if any) so
        the caller can roll back without re-discovering it.

    Raises:
        RuntimeError: when the endpoint has no served entities to update,
            or when the SDK update call fails.
    """
    from databricks.sdk.service.serving import ServedEntityInput

    if ws is None:
        from ._defaults import _make_workspace_client
        ws = _make_workspace_client()

    current = ws.serving_endpoints.get(endpoint_name)
    config = getattr(current, "config", None)
    served_entities = getattr(config, "served_entities", None) if config else None
    if not served_entities:
        raise RuntimeError(
            f"Endpoint {endpoint_name!r} has no served entities to update. "
            "Has the agent been deployed via databricks.agents.deploy yet?"
        )

    previous_value: str | None = None
    new_served_entities: list[ServedEntityInput] = []
    for entity in served_entities:
        existing_env = dict(getattr(entity, "environment_vars", None) or {})
        if previous_value is None and env_var in existing_env:
            previous_value = existing_env[env_var]
        existing_env[env_var] = model

        new_served_entities.append(
            ServedEntityInput(
                name=getattr(entity, "name", None),
                entity_name=getattr(entity, "entity_name", None),
                entity_version=getattr(entity, "entity_version", None),
                workload_size=getattr(entity, "workload_size", None),
                workload_type=getattr(entity, "workload_type", None),
                scale_to_zero_enabled=getattr(entity, "scale_to_zero_enabled", None),
                min_provisioned_concurrency=getattr(entity, "min_provisioned_concurrency", None),
                max_provisioned_concurrency=getattr(entity, "max_provisioned_concurrency", None),
                environment_vars=existing_env,
            )
        )

    if wait:
        ws.serving_endpoints.update_config_and_wait(
            name=endpoint_name,
            served_entities=new_served_entities,
        )
    else:
        ws.serving_endpoints.update_config(
            name=endpoint_name,
            served_entities=new_served_entities,
        )

    return HotSwapResult(
        endpoint_name=endpoint_name,
        new_model=model,
        previous_model=previous_value,
        served_entities_updated=len(new_served_entities),
    )


def get_active_override(
    endpoint_name: str,
    *,
    ws: "WorkspaceClient | None" = None,
    env_var: str = APX_MODEL_OVERRIDE_ENV,
) -> str | None:
    """Return the current value of the model override env var, or None if unset.

    Useful for verifying a swap landed and for showing operators which
    model is actually serving (vs. the compile-time default).
    """
    if ws is None:
        from ._defaults import _make_workspace_client
        ws = _make_workspace_client()

    current = ws.serving_endpoints.get(endpoint_name)
    config = getattr(current, "config", None)
    served_entities = getattr(config, "served_entities", None) if config else None
    if not served_entities:
        return None

    for entity in served_entities:
        env = getattr(entity, "environment_vars", None) or {}
        if env_var in env:
            return env[env_var]
    return None

File: typescript/src/hot-swap.ts
/**
 * hot-swap — change a deployed agent's LLM endpoint without re-logging.
 *
 * The runtime side reads `APX_AGENT_MODEL_OVERRIDE` and uses it instead
 * of the compile-time model. This module is the control side: writes the
 * env var onto a deployed serving endpoint so the next replica picks up
 * the new model.
 *
 * Use cases:
 *   * Swap from `claude-sonnet` to `claude-opus` to test a more
 *     capable model without re-deploying the agent artifact.
 *   * Roll back a problematic model change in seconds (set the override
 *     to the previous value returned by the prior call).
 *
 * For traffic-split A/B testing across two complete artifacts (not just
 * the LLM endpoint), see `canary.ts` instead.
 *
 * Mechanism: rewrites each served entity's `environmentVars` to include
 * the override. Other env vars are preserved. The update is durable
 * until the next `hotSwapModel` call or until the endpoint is redeployed.
 *
 * Caveats:
 *   * The new env var takes effect on the next replica start. With
 *     scale-to-zero, the next request triggers a cold start that picks
 *     it up. Without scale-to-zero, existing replicas may serve the
 *     old model until they cycle.
 */

import { z } from 'zod';
import type {
  ServedEntityInput,
  ServingEndpointsClient,
} from './canary.js';

// ---------------------------------------------------------------------------
// Constants + types
// ---------------------------------------------------------------------------

/**
 * The env var name the framework runtime reads to override the
 * compile-time model. Exposed as a module constant so callers can match
 * it in their own tooling.
 */
export const APX_MODEL_OVERRIDE_ENV = 'APX_AGENT_MODEL_OVERRIDE';

/** Outcome of a hot-swap operation. */
export interface HotSwapResult {
  readonly endpoint: string;
  readonly newModel: string;
  /** The value the override held before; null if first set. */
  readonly previousModel: string | null;
  /** Names of the served entities that were updated. */
  readonly servedEntities: string[];
}

// ---------------------------------------------------------------------------
// hotSwapModel
// ---------------------------------------------------------------------------

const HotSwapModelOpts = z.object({
  endpoint: z.string().min(1),
  model: z.string().min(1),
  servingEndpointsClient: z.custom<ServingEndpointsClient>(
    (v) => typeof v === 'object' && v !== null && 'get' in v && 'updateConfig' in v,
    'servingEndpointsClient must implement get/updateConfig',
  ),
  envVar: z.string().min(1).default(APX_MODEL_OVERRIDE_ENV),
  wait: z.boolean().default(true),
});

/**
 * Hot-swap a deployed agent's LLM endpoint by updating env vars.
 *
 * @returns `HotSwapResult` with the previous override value (if any) so
 *   the caller can roll back without re-discovering it.
 *
 * @throws when the endpoint has no served entities to update.
 */
export async function hotSwapModel(opts: {
  endpoint: string;
  model: string;
  servingEndpointsClient: ServingEndpointsClient;
  envVar?: string;
  wait?: boolean;
}): Promise<HotSwapResult> {
  const { endpoint, model, servingEndpointsClient, envVar } =
    HotSwapModelOpts.parse(opts);

  const info = await servingEndpointsClient.get(endpoint);
  const config = info?.config ?? info?.pendingConfig ?? null;
  const servedEntities = config?.servedEntities ?? config?.servedModels ?? null;

  if (!servedEntities || servedEntities.length === 0) {
    throw new Error(
      `Endpoint '${endpoint}' has no served entities to update. ` +
        'Has the agent been deployed yet?',
    );
  }

  let previousValue: string | null = null;
  const newServedEntities: ServedEntityInput[] = [];
  const updatedNames: string[] = [];

  for (const entity of servedEntities) {
    // Read existing env vars off the response — duck-typed since
    // ServedEntityResponse does not declare `environmentVars`.
    const existing = _readEnvironmentVars(entity);
    const merged: Record<string, string> = { ...existing };

    if (previousValue === null && envVar in merged) {
      previousValue = merged[envVar];
    }
    merged[envVar] = model;

    // Read other fields (also duck-typed; canary's ServedEntityResponse
    // is a subset, but the underlying object likely has more).
    const raw = entity as unknown as Record<string, unknown>;
    const name = _asString(raw['name']);
    const entityName = _asString(raw['entityName'] ?? raw['modelName']);
    const entityVersion = _asString(raw['entityVersion'] ?? raw['modelVersion']);
    const scaleToZero = _asBool(raw['scaleToZeroEnabled']);
    const workloadSize = _asString(raw['workloadSize']);

    newServedEntities.push({
      name: name ?? '',
      entityName: entityName ?? null,
      entityVersion: entityVersion ?? null,
      scaleToZeroEnabled: scaleToZero ?? true,
      workloadSize: workloadSize ?? 'Small',
      environmentVars: merged,
    });
    updatedNames.push(name ?? '');
  }

  await servingEndpointsClient.updateConfig(endpoint, {
    servedEntities: newServedEntities,
  });

  return Object.freeze({
    endpoint,
    newModel: model,
    previousModel: previousValue,
    servedEntities: updatedNames,
  });
}

// ---------------------------------------------------------------------------
// getActiveOverride
// ---------------------------------------------------------------------------

const GetActiveOverrideOpts = z.object({
  endpoint: z.string().min(1),
  servingEndpointsClient: z.custom<ServingEndpointsClient>(
    (v) => typeof v === 'object' && v !== null && 'get' in v && 'updateConfig' in v,
    'servingEndpointsClient must implement get/updateConfig',
  ),
  envVar: z.string().min(1).default(APX_MODEL_OVERRIDE_ENV),
});

/**
 * Return the current value of the model override env var, or null if
 * unset.
 *
 * Useful for verifying a swap landed and for showing operators which
 * model is actually serving (vs. the compile-time default).
 *
 * Mirrors the Python behavior: returns the value from the FIRST served
 * entity that has the env var set, or null if none do (or there are no
 * served entities at all).
 */
export async function getActiveOverride(opts: {
  endpoint: string;
  servingEndpointsClient: ServingEndpointsClient;
  envVar?: string;
}): Promise<string | null> {
  const { endpoint, servingEndpointsClient, envVar } =
    GetActiveOverrideOpts.parse(opts);

  const info = await servingEndpointsClient.get(endpoint);
  const config = info?.config ?? info?.pendingConfig ?? null;
  const servedEntities = config?.servedEntities ?? config?.servedModels ?? null;
  if (!servedEntities || servedEntities.length === 0) {
    return null;
  }

  for (const entity of servedEntities) {
    const env = _readEnvironmentVars(entity);
    if (envVar in env) {
      return env[envVar];
    }
  }
  return null;
}

// ---------------------------------------------------------------------------
// Internal helpers — duck-typed reads off the SDK response.
// ---------------------------------------------------------------------------

function _readEnvironmentVars(entity: unknown): Record<string, string> {
  if (entity == null || typeof entity !== 'object') return {};
  const raw = entity as Record<string, unknown>;
  const env = raw['environmentVars'] ?? raw['environment_vars'];
  if (env && typeof env === 'object' && !Array.isArray(env)) {
    const out: Record<string, string> = {};
    for (const [k, v] of Object.entries(env as Record<string, unknown>)) {
      out[k] = String(v);
    }
    return out;
  }
  return {};
}

function _asString(v: unknown): string | null {
  if (v === null || v === undefined) return null;
  return String(v);
}

function _asBool(v: unknown): boolean | null {
  if (v === null || v === undefined) return null;
  return Boolean(v);
}
